instructor_notes/11_git-init-server.md

```diff
@@ -0,0 +1,116 @@
+# Instructor Notes: Initializing a git repository on RStudio Server
+
+This document provides instructions for creating a git repository in an existing directory on the Data Lab's RStudio Server, as well as setting up associated git credentials.
+
+## Learning goals
+
+At the end of this activity, workshop attendees will have:
+
+* Learned about git Personal Access Tokens and credentials
+* Learned basic git configuration using R tools
+* Used basic git CLI commands
+
+## Part 1: Set up git credentials
+
+### Set up the `~/.gitconfig` file
+
+This can either be done directly in the editor by opening `~/.gitconfig`, or with `usethis` which may be safer for ensuring correct spacing:
+
+In the end, the `~/.gitconfig` file should look like this (where the `init` block is optional but recommended):
+
+```yml
+[user]
+    email = Participant Email
+    name = Participant Name
+[credential]
+    helper = cache --timeout=43200
+[init]
+    defaultBranch = "main"
+```
+
+The `usethis` code to add this content to `~/.gitconfig` is below.
+If `usethis` is used, instruct participants to view the `~/.gitconfig` file after each step to see how the file has been modified.
+Note that `usethis::use_git_config()` will not modify (or delete!) other fields besides those provided to the function.
+
+```r
+# Step 1: Create config file with user and email
+usethis::use_git_config(
+    user.name = "Participant Name",
+    user.email = "Participant Email",
+    # Optional setting to ensure the config is global for the user
+    scope = "user",
+    # Optional setting to use `main` as default branch when creating a new repo
+    init.defaultBranch = "main"
+)
+# Step 2: Set up credential helper to be able to temporarily store PAT on the server
+# Note that this could be combined with step 1 as an additional argument
+usethis::use_git_config(
+    credential.helper = "cache --timeout=43200" # 12 hours
+)
```

Comment:
Is there a reason to do this separately?

Reply:
My thought was setting email/user is going to be a pretty standard thing they do on any system, but this cache timeout may end up being more specific to this workshop, and separating the steps might help to emphasize it (and I should add a comment to explain this more!). What do you think?

Reply:
That makes sense to me.

Reply:
How were you planning to present this? This seems like something to _maybe_ do with slides, but I think we should probably default to one, and the more general solution seems like the better way to go. 

Which also has me wondering if we actually need the `gitcreds` step or if setting the credential helper is enough that when we do a first `git push` the token will be stored? If so, I bet that pasting at the terminal is hidden, which saves some hiding...

Reply:
> Which also has me wondering if we actually need the gitcreds step or if setting the credential helper is enough that when we do a first git push the token will be stored

Without setting the token with `gitcreds`, it prompts you for username and password when pushing, and then tells you you're not allowed to have done that. Without `gitcreds`, I don't know where it would get the token to succeed here to begin with in the first place?

```
stephanie:~/test-repo$ git push
Username for 'https://github.com': sjspielman
Password for 'https://sjspielman@github.com': 
remote: Support for password authentication was removed on August 13, 2021.
remote: Please see https://docs.github.com/get-started/getting-started-with-git/about-remote-repositories#cloning-with-https-urls for information on currently recommended modes of authentication.
fatal: Authentication failed for 'https://github.com/sjspielman/test-repo/'
```

Reply:
> How were you planning to present this? This seems like something to maybe do with slides, but I think we should probably default to one, and the more general solution seems like the better way to go.

The more general solution being `git config`? Totally good with me, I figured one of the approaches would get booted during review! And yes I do plan to put all this in a set of slides before we actually do it live together.

Reply:
Yes `git config` being the more general

Reply:
> Without setting the token with `gitcreds`, it prompts you for username and password when pushing, and then tells you you're not allowed to have done that.

Did you paste in your password or PAT there?

Reply:
GOOD POINT!

With actually pasting PAT, it seems we can indeed skip `gitcreds`. Once the PAT is used, the creds do get populated (`gitcreds::gitcreds_get()` returns a value). Each new terminal session will require them to enter username/PAT again, which doesn't necessarily change the number of times they'd be copy/pasting the PAT in anyways. And indeed, terminal pasting is hidden.

Reply:
> With actually pasting PAT, it seems we can indeed skip `gitcreds`. Once the PAT is used, the creds do get populated (`gitcreds::gitcreds_get()` returns a value). Each new terminal session will require them to enter username/PAT again, which doesn't necessarily change the number of times they'd be copy/pasting the PAT in anyways. And indeed, terminal pasting is hidden.

Let's go ahead and do this then. The more general solution (i.e. not requiring R) is going to be my preference here too. Can you update these instructions accordingly?



> Totally good with me, I figured one of the approaches would get booted during review!

I think it is okay if both live in this doc, but there should be clear instructions about what we will present.